StackEng/Opserver
# ASCN-377:  Fix for certain features not working due to SQLClient bugs

# The Issue
When running in a docker container, certain features of opserver fail:
- Go to https://opserver.main-stg.int.gcp.stackoverflow.net/sql/databases?node=db.db
- Click on a Database that has data (e.g. Sites)
- You will see a 500 error
- The logs will show: `Internal connection fatal error.` in `System.Data.SqlClient.TdsParser.TryRun(RunBehavior runBehavior, SqlCommand cmdHandler, SqlDataReader dataStream, BulkCopySimpleResultSet bulkCopyHandler, TdsParserStateObject stateObj, Boolean& dataReady)`
- This seems related to this reported issue:  https://github.com/dotnet/SqlClient/issues/365
- You can also duplicate this locally in docker.  If you install opserver in docker, you'll have the same issue but won't if you test outside of docker

# The solution
Switching to the not-out-of-support `Microsoft.Data.SqlClient` resolves this issue.  Unfortunately that causes these side-effects:

## invariant fixes
Using this client, we have to switch to the chiseled extra base image to get around known issues with global invariant. Without the change the app won't connect to SQL

## Certificate not trusted issue
Without adding `;TrustServerCertificate=True` the app won't connect to sql.  I believe we saw this in other apps?

## Microsoft.IdentityModel.Protocols.OpenIdConnect
The updated `Microsoft.Data.SqlClient` uses an updated reference to this library.  We have to explicitly upgrade it in the project

## Exceptional still uses System.Data.SqlClient
Although our code no longer uses it, Exceptional still needs this.  The app won't start without it being referenced at least in the web project.


# How to test
Locally:
- Set your `opserversettings.json` -- uncomment the stuff on line 10-14;  add `TrustServerCertificate=True` to the connection string
- Add a section before `Modules` -- `Security: { "Provider": "EveryonesAnAdmin"},`
- Run within VS - make sure everything works
- Run `.\cnab\invoke-CNAB.ps1 install $false DockerDesktop` -- **Make sure you get a new container by looking at the pod in lens. If you don't, delete the pod so it recreates after install**
